test/he/test_pmtud.c

```diff
@@ -0,0 +1,269 @@
+  // Should be using big step when first entering searching state
+  TEST_ASSERT_TRUE(conn.pmtud_is_using_big_step);
+  // Current probing size should be BASE_PMTU + BIG_STEP
+  TEST_ASSERT_EQUAL(conn.pmtud_base + PMTUD_PROBE_BIG_STEP, conn.pmtud_probing_size);
+  // Both callbacks should be called
+  TEST_ASSERT_EQUAL(2, call_counter);
+  // Probe count should be reset
+  TEST_ASSERT_EQUAL(1, conn.pmtud_probe_count);
+}
+
+void test_he_internal_pmtud_handle_probe_ack_from_error(void) {
+  conn.state = HE_STATE_ONLINE;
+  conn.pmtud_state = HE_PMTUD_STATE_ERROR;
+  conn.pmtud_probe_pending_id = 123;
+  conn.pmtud_state_change_cb = pmtud_state_change_cb;
+  conn.pmtud_time_cb = pmtud_time_cb;
+  conn.pmtud_base = HE_MAX_MTU;
+
+  EXPECT_HE_INTERNAL_SEND_MESSAGE();
+
+  TEST_ASSERT_EQUAL(HE_SUCCESS, he_internal_pmtud_handle_probe_ack(&conn, 123));
+
+  // New state should be searching
+  TEST_ASSERT_EQUAL(HE_PMTUD_STATE_SEARCHING, conn.pmtud_state);
+  // Should be using big step when first entering searching state
+  TEST_ASSERT_TRUE(conn.pmtud_is_using_big_step);
+  // Current probing size should be BASE_PMTU + BIG_STEP
+  TEST_ASSERT_EQUAL(conn.pmtud_base + PMTUD_PROBE_BIG_STEP, conn.pmtud_probing_size);
+  // Both callbacks should be called
+  TEST_ASSERT_EQUAL(2, call_counter);
+  // Probe count should be reset
+  TEST_ASSERT_EQUAL(1, conn.pmtud_probe_count);
+}
+
+static void test_handle_probe_ack_from_searching(uint16_t probe_size, bool use_big_step) {
+  conn.state = HE_STATE_ONLINE;
+  conn.pmtud_state = HE_PMTUD_STATE_SEARCHING;
+  conn.pmtud_probe_pending_id = 123;
+  conn.pmtud_state_change_cb = pmtud_state_change_cb;
+  conn.pmtud_time_cb = pmtud_time_cb;
+  conn.pmtud_base = MIN_PLPMTU;
+  conn.pmtud_probing_size = probe_size;
+  conn.pmtud_is_using_big_step = use_big_step;
+
+  if(probe_size < MAX_PLPMTU) {
+    EXPECT_HE_INTERNAL_SEND_MESSAGE();
+  }
+
+  TEST_ASSERT_EQUAL(HE_SUCCESS, he_internal_pmtud_handle_probe_ack(&conn, 123));
+
+  if(probe_size >= MAX_PLPMTU) {
+    // State should enter search complete
+    TEST_ASSERT_EQUAL(HE_PMTUD_STATE_SEARCH_COMPLETE, conn.pmtud_state);
+    // Effective PMTU should be set
+    TEST_ASSERT_EQUAL(MAX_PLPMTU, conn.effective_pmtu);
+    // pmtud_state_change_cb callback should be called
+    TEST_ASSERT_EQUAL(1, call_counter);
+  } else {
+    // State should still be searching if the probed size is smaller than MAX_PLPMTU
+    TEST_ASSERT_EQUAL(HE_PMTUD_STATE_SEARCHING, conn.pmtud_state);
+    // Probe size should be incremented
+    if(use_big_step) {
+      TEST_ASSERT_EQUAL(probe_size + PMTUD_PROBE_BIG_STEP, conn.pmtud_probing_size);
+    } else {
+      TEST_ASSERT_EQUAL(probe_size + PMTUD_PROBE_SMALL_STEP, conn.pmtud_probing_size);
+    }
+    // Only the pmtud_time_cb callback should be called
+    TEST_ASSERT_EQUAL(1, call_counter);
+  }
+}
+
+void test_he_internal_pmtud_handle_probe_ack_from_searching_big_step(void) {
+  test_handle_probe_ack_from_searching(1212, true);
```

Comment:
Maybe we should define this as `MAX_PLPMTU - some value` instead of `1212`, so that:
1. It's clear that this value is less than MAX_PLPMTU, as that's critical to one of the if statements in the helper function
2. If we re-define MAX_PLPMTU to be less than 1212 in the future, this test doesn't break / fail with a false positive.